lib/optimizer/src/Transformer/PreloadHeroImage.php

```diff
@@ -0,0 +1,563 @@
+                $node = $this->nextNode($node);
+                continue;
+            }
+
+            $heroImage = $this->detectImageWithDataHero($node);
+            if ($heroImage) {
+                $heroImages[] = $heroImage;
+            }
+
+            if (! $heroImageCandidate && count($heroImages) === 0) {
+                $heroImageCandidate = $this->detectHeroImageCandidate($node);
+            }
+            if (Amp::isTemplate($node)) {
+                // Ignore images inside templates.
+                $node = $this->skipNodeAndChildren($node);
+            } else {
+                $node = $this->nextNode($node);
+            }
+        }
+
+        // Optimize data-hero element if defined.
+        if (count($heroImages) > 0) {
+            return $heroImages;
+        }
+
+        // Fall back to auto-detected hero image if available.
+        if ($heroImageCandidate) {
+            return [$heroImageCandidate];
+        }
+
+        // No hero images to optimize.
+        return [];
+    }
+
+    /**
+     * Detect a hero image with the data-hero attribute.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image, or null if none detected.
+     */
+    private function detectImageWithDataHero(DOMElement $element)
+    {
+        if (
+            $element->tagName === Extension::IMAGE
+            && $element->hasAttribute(Attribute::DATA_HERO)
+        ) {
+            return new HeroImage(
+                $element->getAttribute(Attribute::SRC),
+                $element->getAttribute(Attribute::MEDIA),
+                $element->getAttribute(Attribute::SRCSET),
+                $element
+            );
+        }
+
+        if (
+            Amp::isAmpIframe($element)
+            && $element->hasAttribute(Attribute::DATA_HERO)
+        ) {
+            return $this->getPlaceholderImage($element);
+        }
+
+        return null;
+    }
+
+    /**
+     * Detect a hero image candidate.
+     *
+     * The hero image here can come from one of <amp-img>, <amp-video>, <amp-iframe>, <amp-video-iframe>.
+     *
+     * @param DOMElement $element Element to detect for.
+     * @return HeroImage|null Detected hero image candidate, or null if none detected.
+     */
+    private function detectHeroImageCandidate(DOMElement $element)
+    {
+        if (
+            $element->hasAttribute(Attribute::LAYOUT)
+            && $element->getAttribute(Attribute::LAYOUT) === Layout::NODISPLAY
+        ) {
+            return null;
+        }
+
+        if ($element->tagName === Extension::IMAGE) {
+            return $this->detectHeroImageCandidateForAmpImg($element);
+        }
+
+        if ($element->tagName === EXTENSION::VIDEO) {
+            return $this->detectHeroImageCandidateForPosterImage($element);
+        }
+
+        if (Amp::isAmpIframe($element)) {
+            return $this->detectHeroImageCandidateForIframePlaceholderImage($element);
+        }
```

Comment:
We could also detect a hero (poster) image for `amp-youtube`. The AMP plugin puts an `amp-img[placeholder]` pointing to the `hqdefault.jpg` for the video.

Reply:
According to [the docs](https://amp.dev/documentation/components/amp-youtube/?format=websites), the `<amp-youtube>` component does not support a `poster` attribute. It does support an `<amp-img>` child element that is marked as `placeholder`, so that could indeed be added.

Reply:
Yes, that's what I'm referring to:

https://github.com/ampproject/amp-wp/blob/c39f672a6361c99be7cab5cd58794cfebc7cd4d4/tests/php/test-class-amp-youtube-embed-handler.php#L151

Reply:
Apparently to support this, the `detectHeroImageCandidateForIframePlaceholderImage` method can just be renamed to `detectHeroImageCandidateForPlaceholderImage`, and instead of looking for direct descendants of the element for image placeholders, it should look for a `placeholder` direct descendant and then check if either it is an `amp-img` _or_ it has an `amp-img` as a descendant (as in the case of a placeholder link having an image inside).